PLT-5717 Initial markdown support

hmhealey opened the pull request:
What this includes:
- Bold/Italics
- Explicit links `[Google](http://google.ca)`
- Headings
- Code blocks (fenced and indented)
- Inline code
- Lists
- Horizontal ines

------

What this doesn't include (https://mattermost.atlassian.net/browse/PLT-5736):
- ~~Strikethrough text~~ (not part of commonmark spec, https://mattermost.atlassian.net/browse/PLT-5736)
- Tables (not part of commonmark spec, https://mattermost.atlassian.net/browse/PLT-5736)
- Images (ran into implementation difficulties, https://mattermost.atlassian.net/browse/PLT-5736)
- Emoji (https://mattermost.atlassian.net/browse/PLT-5537)
- Various mentions and hashtags

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-5717

#### Checklist
- Has UI changes

#### Device Information
This PR was tested on: Nexus 5 (Android 6.0.1) and iPhone Emulator (iPhone 6, iOS 10.1 )

#### Screenshots
![image](https://cloud.githubusercontent.com/assets/3277310/23960406/4230b0e4-097e-11e7-8660-628361e1b4aa.png)

![image](https://cloud.githubusercontent.com/assets/3277310/23960360/33cca364-097e-11e7-9878-38720b119753.png)

enahum commented on app/components/markdown/markdown_link.js:
```diff
@@ -0,0 +1,28 @@
+    handlePress = () => {
+        const url = this.props.href;
+
+        Linking.canOpenURL(url).then((supported) => {
```
Nothing to do here I think this is ok for now, but we should think about url's like permalinks and how are we going to handle them

hmhealey replied:
That'll be easy enough. We already look at the URL in the webapp to figure out how to handle the link so we can do something similar here

enahum commented on app/components/post/post.js:
```diff
@@ -117,23 +118,35 @@ export default class Post extends Component {
-                        {this.props.post.message}
-                    </Text>
-                    }
+                    {contents}
```
Nothing to do here, but c'mon this will conflict too often hehehe

enahum commented on app/constants/custom_prop_types.js:
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) 2017 Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+import {PropTypes} from 'react';
```
Can we find a better place for this other than `constants` folder?

hmhealey replied:
How about `utils`? I was going back and forth between both of them

enahum replied:
yeah `utils` sounds a little bit better

enahum commented on package.json:
```diff
@@ -3,6 +3,8 @@
   "private": true,
   "dependencies": {
+    "commonmark": "0.27.0",
+    "commonmark-react-renderer": "hmhealey/commonmark-react-renderer#55da1c569c497db7b58d2f4d8c5e6c2fbbf3f2ff",
```
can we move this library to `mattermost` org?

hmhealey replied:
Yep, I was going to do that as well, but I was waiting until I fork commonmark as well

enahum replied:
alright we can do this at any time, actually I have to do the same with `react-native-search-bar`

enahum commented:
@hmhealey when testing I found these
PS: only tested on IOS

1. image links go outside of the view
![img_0323](https://cloud.githubusercontent.com/assets/6757047/23999767/0ba9517e-0a38-11e7-994d-7d5b2be9bab5.PNG)

2. List do not ident
![img_0324](https://cloud.githubusercontent.com/assets/6757047/23999789/17a82892-0a38-11e7-9c6e-2b02c963f1f0.PNG)

3.  inline `code` hardly readable in dark theme
![img_0326](https://cloud.githubusercontent.com/assets/6757047/23999858/585b6994-0a38-11e7-9a70-62eff16c7fbc.PNG)

4. code blocks do not scroll
![img_0327](https://cloud.githubusercontent.com/assets/6757047/23999866/616770be-0a38-11e7-9bb4-b235267eb610.PNG)

hmhealey commented:
1. I thought I had that working, but it must've broken on me. I'm going to switch that back to just generating a placeholder since images aren't as high priority as other features. I think I'm going to need to do some more work for images to hoist them out of the surrounding text since they're technically inline right now.
2. That's actually intended by CommonMark (and some flavours of Markdown). Eric's post mixes bullet points which marked.js treats as a single list, but others consider as two separate lists. It looks like this in the text:
```
1. Fixed an issue...
- Fixed another issue...
```
3. Yep, I'll fix that. The text colour isn't being applied inside the code
4. It is actually in a ScrollView, but the pan handler for the menu kept intercepting my input to scroll it. I'm not sure if there's anything we can do apart from making the pan handler less sensitive.

hmhealey commented:
^ @enahum

enahum commented:
@hmhealey yes lets see if we can make the pan handler for the drawers a little bit smaller to see if that helps

enahum commented:
@hmhealey yes lets see if we can make the pan handler for the drawers a little bit smaller to see if that helps

enahum commented:
@hmhealey I'll test this out again by the end of the day
